### PR TITLE
Update validator to handle new answer_count routing

### PR DESCRIPTION
**Pull request by insacuri**
https://trello.com/c/sBtFoRAQ/2482-add-routing-on-the-count-of-multiple-answers-to-when-rule-s

### CODE REVIEW

**Comment by insacuri** [comment]
@MebinAbraham Added duplicate check

File: schemas/common_definitions.json [hunk 1]
```diff
@@ -193,6 +201,11 @@
           "required": [
             "answer_count"
```
Comment (MebinAbraham) [inline schemas/common_definitions.json]: Pretty sure this is wrong?. This is valid if `oneOf` them is provided but we need more than 1.

File: app/validation/validator.py [hunk 2]
```diff
@@ -302,25 +302,29 @@ def validate_when_rule(when_clause, answer_ids_with_group_id, referenced_id):
         Validates any answer id in a when clause exists within the schema
         Will also check that comparison_id exists
         """
-        answer_reference_id_fields = ('id', 'answer_count', 'comparison_id')
+        answer_reference_id_fields = ('id', 'answer_ids', 'answer_count', 'comparison_id')
```
Comment (ajmaddaford) [inline app/validation/validator.py]: Isn't `answer_count` no longer a property that can exist in a when rule? (as it's now a value of a `type` property)

File: schemas/common_definitions.json [hunk 3]
```diff
@@ -118,6 +118,14 @@
           "type": "string",
           "description": "A repeated group id, for which the count will be fetched."
         },
+        "type": {
+          "type": "string",
+          "description": "Type of rule used for when condition."
```
Comment (ajmaddaford) [inline schemas/common_definitions.json]: This should be an enum containing "answer_count"

File: schemas/common_definitions.json [hunk 4]
```diff
@@ -193,6 +201,11 @@
           "required": [
             "answer_count"
           ]
+        },
+        {
+          "required": [
+            "type"
```
Comment (ajmaddaford) [inline schemas/common_definitions.json]: `type` can't be required as most when rules don't define a `type`

File: app/validation/validator.py [hunk 5]
```diff
@@ -302,25 +302,29 @@ def validate_when_rule(when_clause, answer_ids_with_group_id, referenced_id):
...
+                    if when[id_ref_key] not in answer_ids_with_group_id:
+                        return Validator._error_message('The answer id - {} in the {} key of the "when" clause for {} does not exist'
+                                                        .format(when[id_ref_key], id_ref_key, referenced_id))
 
-                if when[id_ref_key] not in answer_ids_with_group_id:
-                    return Validator._error_message('The answer id - {} in the {} key of the "when" clause for {} does not exist'
-                                                    .format(when[id_ref_key], id_ref_key, referenced_id))
-
-                if id_ref_key == 'answer_count':
+                if id_ref_key == 'answer_count' or when.get('type') == 'answer_count':
```
Comment (ajmaddaford) [inline app/validation/validator.py]: This should be moved out of the loop - it doesn't make sense in the loop of `present_id_ref_keys` as the when type does not contain an answer id.

File: app/validation/validator.py [hunk 6]
```diff
@@ -333,6 +337,9 @@ def validate_when_rule(when_clause, answer_ids_with_group_id, referenced_id):
                     return Validator._error_message('The "when" clause for {} contains a comparison_id and uses a '
                                                     'condition of unset or set'
                                                     .format(referenced_id))
+            if 'answer_ids' in when:
+                if len(when['answer_ids']) != len(set(when['answer_ids'])):
+                    return Validator._error_message('Duplicate answer ids found within {} clause'.format(referenced_id))
```
Comment (ajmaddaford) [inline app/validation/validator.py]: I realise this is existing, but I can't see why it returns once it's found one error. The other validation methods build an array of errors and return the array.

File: schemas/common_definitions.json (hunk 4 above)
Comment (insacuri) [reply schemas/common_definitions.json]: Isn't this oneOf for the whole block?  I thought it was saying 'you need one of id, meta, answer_count, type', is that wrong?

File: schemas/common_definitions.json (hunk 3 above)
Comment (insacuri) [reply schemas/common_definitions.json]: Done

File: schemas/common_definitions.json (hunk 4 above)
Comment (ajmaddaford) [reply schemas/common_definitions.json]: My bad, that'll teach me for commenting on github without viewing the whole file 🤦‍♂️

File: app/validation/validator.py (hunk 6 above)
Comment (insacuri) [reply app/validation/validator.py]: Fixed

Comment (insacuri) [reply app/validation/validator.py]: Fixed

**Pull request merged**